include/backend/backend.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include "async_event.h"
+#include "closure_hydrator.h"
+#include "launcher.h"
+#include "types.h"
+
+#include <vector>
+
+#include <sycl/sycl.hpp>
+
+namespace celerity::detail {
+
+class communicator;
+struct system_info;
+
+/// The backend is responsible for asynchronously allocating device- and device-accessible host memory, copying data between host and device allocations, and
+/// launching host tasks and device kernels. Asynchronous work must be explicitly assigned to in-order queue ids as assigned by `out_of_order_engine`.
+class backend {
+  public:
+	backend() = default;
+	backend(const backend&) = delete;
+	backend(backend&&) = delete;
+	backend& operator=(const backend&) = delete;
+	backend& operator=(backend&&) = delete;
+	virtual ~backend() = default;
+
+	/// Returns metadata about the system as it appears to the backend implementation.
+	virtual const system_info& get_system_info() const = 0;
+
+	/// Synchronously allocates device-accessible host memory. This is slow and meant for debugging purposes only.
+	virtual void* debug_alloc(size_t size) = 0;
+
+	/// Synchronously frees device-accessible host memory. This is slow and meant for debugging purposes only.
+	virtual void debug_free(void* ptr) = 0;
+
+	/// Schedules the allocation of device-accessible host memory with the specified size and alignment. The operation will complete in-order with respect to
+	/// any other asynchronous `alloc` and `free` operation on the same backend.
+	virtual async_event enqueue_host_alloc(size_t size, size_t alignment) = 0;
+
+	/// Schedules the allocation of device memory with the specified size and alignment. The operation will complete in-order with respect to any other
+	/// asynchronous `alloc` and `free` operation on the same backend.
+	virtual async_event enqueue_device_alloc(device_id memory_device, size_t size, size_t alignment) = 0;
```

Comment:
Yes. `memory_device` is not part of a lane identifier, there is only a single lane, because at least CUDA serializes all device-accessible memory allocation anyway.